lua/autorun/server/sv_chip_lister.lua

```diff
@@ -202,116 +200,123 @@ local function prepareChipName( str )
+        if not plyData then
+            plyData = {
+                Count = 0,
+                OwnerIndex = owner == ID_WORLD and 0 or owner:EntIndex(),
+                OwnerName = owner == ID_WORLD and ID_WORLD or preparePlyName( getNick( owner ) ),
+                OwnerTotalUsage = 0,
+                ChipInfo = {},
+            }
+            playerData[owner] = plyData
+            lineCount = lineCount + 1
+        end
 
-    -- For some reason, :IsPlayer() always returns false if obtained locally from entityMeta, it HAS to be called this way
-    if IsValid( owner ) and owner:IsPlayer() then
-        ownerName = preparePlyName( getNick( owner ) )
-    else
-        owner = ID_WORLD
-        ownerName = ID_WORLD
+        table.insert( plyData.ChipInfo, {
+            Name = prepareChipName( getChipName( chip ) ),
+            IsE2 = isE2,
+            CPUUsage = chipUsage == false and -1 or chipNormalizedUsage,
+        } )
+        plyData.Count = plyData.Count + 1
+        if chipUsage ~= false then
+            plyData.OwnerTotalUsage = plyData.OwnerTotalUsage + chipNormalizedUsage
+        end
+        lineCount = lineCount + 1
     end
 
-    local id = rawget( idLookup, owner )
-
-    if not id then
-        if elemCount > MAX_TOTAL_ELEMENTS then return globalUsage, idCount, elemCount end
-
-        idCount = idCount + 1
-        id = idCount
-        elemCount = elemCount + 1
-        rawset( idLookup, owner, id )
+    local sortedPlayerData = {}
+    for _, data in pairs( playerData ) do
+        table.insert( sortedPlayerData, data )
     end
+    table.sort( sortedPlayerData, function( a, b )
+        return a.OwnerTotalUsage > b.OwnerTotalUsage
+    end )
 
-    local data = rawget( perPlyData, id )
-    local dataCount
-
-    if elemCount > MAX_TOTAL_ELEMENTS then
-        if data then
-            rawset( data, "OwnerUsage", rawget( data, "OwnerUsage" ) + chipUsage )
+    -- If we're over the limit, remove chips from the end of the list until we're under the limit.
+    if lineCount >= MAX_TOTAL_ELEMENTS then
+        local excessLines = lineCount - MAX_TOTAL_ELEMENTS
+        local removedLines = 0
+
+        for _, data in ipairs( sortedPlayerData ) do
+            if data.Count > 0 then
+                local chipCount = #data.ChipInfo
+                if chipCount > 0 then
+                    local chipsToRemove = math.min( chipCount, excessLines - removedLines )
+                    for _ = 1, chipsToRemove do
+                        tableRemove( data.ChipInfo, #data.ChipInfo )
+                    end
+                    data.Count = data.Count - chipsToRemove
+                    removedLines = removedLines + chipsToRemove
+
+                    if data.Count == 0 then
+                        tableRemove( sortedPlayerData, i )
+                        removedLines = removedLines + 1
+                    end
+
+                    if removedLines >= excessLines then break end
+                end
+            end
```

Comment:
```suggestion
        for i = #sortedPlayerData, 1, -1 do
            local data = sortedPlayerData[i]
            local chipInfo = data.ChipInfo
            local chipsToRemove = math.min( #chipInfo, excessLines - removedLines )

            for _ = 1, chipsToRemove do
                chipInfo[#chipInfo] = nil
            end

            data.Count = #chipInfo
            removedLines = removedLines + chipsToRemove

            if #chipInfo == 0 then
                tableRemove( sortedPlayerData, i )
                removedLines = removedLines + 1
            end

            if removedLines >= excessLines then break end
```

Reply:
Tiny perf improvement, fixes a nil error, fixes removal inside of ipairs(), and removes from the end of the list to respect order priority.